src/entities/Area.js

```diff
@@ -58,33 +58,85 @@ export default class Area extends _EntityWithInventory {
-        const entityIDs = player.getAllEntities() || [];
-        if (shouldAdvance && !this.dialogue.isBattle() && this.dialogue.requiresItem() && !this.dialogue.testItems(entityIDs)) {
-          shouldAdvance = false;
-          respond(strings.missingSomething);
-        }
-        // if we've made it this far, the user has entered the correct command or won the
-        // battle, and if an item is required... they've got it! advance the conversation.
-        if (shouldAdvance) {
-          const drops = this.dialogue.getDrops();
-          this.dialogue.advanceConversation();
-          // if the dialogue is now complete, and there are drop items, add them
-          // to our inventory and notify the user
-          if (drops && Object.keys(drops).length) {
-            for (const itemID in drops) {
-              if (drops.hasOwnProperty(itemID)) {
-                this.addEntity(itemID, drops[itemID]);
-                respond(S(strings.fallToFloor).template({itemName: this.entityManager.get(itemID).name, itemCount: drops[itemID]}).s);
+        console.log("dialogue not complete");
+        // 1) Check if the current text block is done (battle, etc.)
+        let executionFinished = this.dialogue.execute(input, respond, player);
+
+        // 2) Do regex stuff to find the next block
+        if (executionFinished) {
+          console.log("executionFinished");
+          let nextTextKey = this.dialogue.getNextTextKey(input);
+
+          if (nextTextKey)
+          {
+            console.log("getting next text key");
+            // Make sure you meet any requirements to enter the new text block
+            const entityIDs = player.getAllEntities() || [];
+            if (this.dialogue.meetsRequirements(nextTextKey, player)) {
+
+              // Advance the dialogue
+              let gotoSuccess = this.dialogue.goToTextState(nextTextKey);
+
+              // if the dialogue is now complete, and there are drop items, add them
+              // to our inventory and notify the user
+              if (gotoSuccess) {
+                const drops = this.dialogue.getDrops();
+                if (drops && Object.keys(drops).length) {
+                  for (const itemID in drops) {
+                    if (drops.hasOwnProperty(itemID)) {
+                      this.addEntity(itemID, drops[itemID]);
+                      respond(S(strings.fallToFloor).template({itemName: this.entityManager.get(itemID).name, itemCount: drops[itemID]}).s);
+                    }
+                  }
+                }
+                respond(this.dialogue.activate());
               }
+              else {
+                // Getting here means we tried to goto a text state that doesn't exist in the conversation
+                // AKA mismatched key, check for typos in JSON.
+                // If Jordan ever makes a JSON form for Patrick this should cease to be an issue
```

Comment:
yeah it would be dope if we made some kind of "Creator" thing that poops out the JSON files